lib/matplotlib/cbook/__init__.py

```diff
@@ -1981,6 +2024,32 @@ def __call__(self, key):
             break
 
 
+def contiguous_regions(mask):
+    """
+    Return a list of (ind0, ind1) such that mask[ind0:ind1].all() is
+    True and we cover all such regions
+    """
+    mask = np.asarray(mask, dtype=bool)
+
+    if not mask.size:
+        return []
+
+    # Find the indices of region changes, and correct offset
+    idx, = np.nonzero(mask[:-1] != mask[1:])
+    idx += 1
+
+    # List operations are faster for moderately sized arrays
+    idx = idx.tolist()
+
+    # Add first and/or last index if needed
+    if mask[0]:
+        idx = [0] + idx
+    if mask[-1]:
+        idx.append(len(mask))
+
+    return list(zip(idx[::2], idx[1::2]))
```

Comment:
I bet the fastest (not that it necessarily matters) is either to return an iterator using itertools.chain and itertools.islice, or to preallocate the result as an appropriately sized array.

Reply:
This is a function I've just directly copied across from mlab to cbook because it's used in 2 different places in Matplotlib

(sorry, I'm kind of dumping stuff as I work on it here, will remove the WIP bit of the title when I think it's all done)

Reply:
no worries, I'm also just dumping my thoughts here and there :)

Reply:
Couldn't get my head around how to make this work so I think I'll leave it for this PR.

Reply:
`np.column_stack([idxs[::2], idxs[1::2]])` should work well.